File: booking-service/src/main/java/dev/project/booking/integration/kafka/event/PaymentEvent.java
package dev.project.booking.integration.kafka.event;

import java.time.Instant;
import java.util.UUID;

public record PaymentEvent(
        UUID eventId,
        String eventType,
        int eventVersion,
        Instant occurredAt,
        String producer,
        String aggregateType,
        UUID aggregateId,
        UUID correlationId,
        Payload payload
) {

    public record Payload(
            UUID paymentId,
            UUID bookingId,
            long amountMinor,
            String currency,
            String status,
            Instant paidAt,
            String failureReason
    ) {
    }

}

File: payment-service/internal/service/payment_event.go
package service

import (
	"payment-service/internal/domain"
	"payment-service/internal/kafka"
	"time"

	"github.com/google/uuid"
)

func buildPaymentEvent(
	payment *domain.Payment,
	eventType string,
) kafka.PaymentEvent {
	return kafka.PaymentEvent{
		EventID:       uuid.New(),
		EventType:     eventType,
		EventVersion:  1,
		OccurredAt:    time.Now().UTC(),
		Producer:      "payment-service",
		AggregateType: "payment",
		AggregateID:   payment.ID,
		CorrelationID: payment.BookingID,
		Payload: kafka.PaymentPayload{
			PaymentID:     payment.ID,
			BookingID:     payment.BookingID,
			AmountMinor:   payment.AmountMinor,
			Currency:      payment.Currency,
			Status:        string(payment.Status),
			PaidAt:        payment.PaidAt,
			FailureReason: payment.FailureReason,
		},
	}
}
